### DISCUSSION TITLE
Improve MongoDB join performance: move $sort and $limit before $lookup in generated aggregation

### DISCUSSION THREAD

**Discussion by timurvafin** [payloadcms/payload#10165] [Feature Requests & Ideas]
When using a "join" field type in Payload CMS (for example, referencing "documents" from "persons"), the generated MongoDB aggregation pipeline places $lookup before $sort, $skip, and $limit. This causes a full collection scan (COLLSCAN) in MongoDB, making queries very slow for large collections.

**Current Behavior**

Payload CMS generates a pipeline like:

```
[
  {
    $lookup: {
      as: "documents.docs",
      from: "documents",
      localField: "_id",
      foreignField: "persons",
      pipeline: [
        { $sort: { createdAt: NumberInt(-1) } },
        { $limit: NumberInt(11) }
      ]
    }
  },
  {
    $addFields: {
      "documents.docs": {
        $map: {
          as: "doc",
          in: "$$doc._id",
          input: "$documents.docs"
        }
      },
      "documents.hasNextPage": {
        $gt: [
          { $size: "$documents.docs" },
          NumberInt(10)
        ]
      }
    }
  },

  { $sort: { createdAt: 1 } },
  { $skip: NumberInt(0) },
  { $limit: NumberInt(10) },
  
]

```

Because $lookup occurs first, MongoDB has to scan the entire collection before sorting and limiting, which often results in a COLLSCAN.

**Why It Is a Problem:**
- The entire "persons" collection is scanned, even if only 10 documents are ultimately needed.
- MongoDB indexes for sorting (e.g., on createdAt) are underutilized, leading to high resource usage and poor performance.
- For large datasets, response times can become prohibitively long.

**Expected/Desired Behavior**

The pipeline should apply $sort, $skip, and $limit before $lookup. For example:

```
[
  { $sort: { createdAt: 1 } },
  { $skip: NumberInt(0) },
  { $limit: NumberInt(10) },

  
  {
    $lookup: {
      as: "documents.docs",
      from: "documents",
      localField: "_id",
      foreignField: "persons",
      pipeline: [
        { $sort: { createdAt: NumberInt(-1) } },
        { $limit: NumberInt(11) }
      ]
    }
  },
  {
    $addFields: {
      "documents.docs": {
        $map: {
          as: "doc",
          in: "$$doc._id",
          input: "$documents.docs"
        }
      },
      "documents.hasNextPage": {
        $gt: [
          { $size: "$documents.docs" },
          NumberInt(10)
        ]
      }
    }
  },  
]
```

In this scenario, only the required 10 "persons" documents are fetched (utilizing an index on createdAt, if present), and then we do $lookup to retrieve related "documents." This avoids a full collection scan and dramatically improves performance.

**Steps to Reproduce**
1.	Create two collections in Payload CMS, e.g. “persons” and “documents”.
2.	In “documents”, store references to “persons” in a field (array or single).
3.	In “persons”, define a “join” field referencing “documents.”
4.	Insert many documents into “persons” (e.g., tens of thousands).
5.	Query “persons” with the join field; observe slow response times and a COLLSCAN in the MongoDB query plan.

**Proposed Solution**

- Change the order of stages in the generated aggregation. Place $sort, $skip, and $limit at the start of the pipeline, then perform $lookup.
- Alternatively, offer an option/flag in Payload to control the order of $lookup vs. $sort/$limit stages in the pipeline.

**Additional Context**

- Payload CMS Version: 3.11.0
- MongoDB Version:  docker mongo:latest

**Thank you!**

I appreciate your work on Payload CMS. If there is anything else I can provide (logs, a reproducible test repo, etc.), please let me know. I’m also open to creating a PR if that would help!

**Comment by r1tsuu** [comment]
This is a good catch
I'm pretty sure I handled this like you said here which includes also other optimizations - https://github.com/payloadcms/payload/pull/9594 https://github.com/payloadcms/payload/pull/9594/files#diff-9ea27c536aac91975dd667ad213cf324c1b127a71ee9a2d269c8d4b820fbda99R47-R71 unfortunately we had to revert it for now
We'll come back to it!

**Reply by timurvafin** [reply]
Hey @r1tsuu, is there any progress on this change, probably I've missed related pr or its still in progress?

**Comment by tsemachh** [comment]
Not sure if it's related I am facing performance issues in the Admin List View where I have 3 join fields.
Tried to :
`disableListColumn: true,
        disableListFilter: true,`
But still the query still produces:
**aggregate with all 3 joins values, relations ,meetings**
`
stocks.aggregate([ { '$facet': { docs: [ { '$sort': { createdAt: -1 } }, { '$match': {} }, { '$lookup': { as: 'values.docs', foreignField: 'stockId', from: 'stock-values', localField: '_id', pipeline: [ { '$match': {} }, { '$sort': { createdAt: -1 } }, { '$limit': 11 } ] } }, { '$addFields': { 'values.docs': { '$map': { as: 'doc', in: '$$doc._id', input: '$values.docs' } }, 'values.hasNextPage': { '$gt': [ { '$size': '$values.docs' }, 10 ] } } }, { '$addFields': { 'values.docs': { '$slice': [ '$values.docs', 10 ] } } }, { '$lookup': { as: 'relations.docs', foreignField: 'relatedStock', from: 'stocks', localField: '_id', pipeline: [ { '$match': {} }, { '$sort': { createdAt: -1 } }, { '$limit': 11 } ] } }, { '$addFields': { 'relations.docs': { '$map': { as: 'doc', in: '$$doc._id', input: '$relations.docs' } }, 'relations.hasNextPage': { '$gt': [ { '$size': '$relations.docs' }, 10 ] } } }, { '$addFields': { 'relations.docs': { '$slice': [ '$relations.docs', 10 ] } } }, { '$lookup': { as: 'meetings.docs', foreignField: 'stock', from: 'meetings', localField: '_id', pipeline: [ { '$match': {} }, { '$sort': { createdAt: -1 } }, { '$limit': 11 } ] } }, { '$addFields': { 'meetings.docs': { '$map': { as: 'doc', in: '$$doc._id', input: '$meetings.docs' } }, 'meetings.hasNextPage': { '$gt': [ { '$size': '$meetings.docs' }, 10 ] } } }, { '$addFields': { 'meetings.docs': { '$slice': [ '$meetings.docs', 10 ] } } }, { '$sort': { stockId: 1 } }, { '$skip': 10 }, { '$limit': 10 } ], count: [ { '$sort': { createdAt: -1 } }, { '$match': {} }, { '$lookup': { as: 'values.docs', foreignField: 'stockId', from: 'stock-values', localField: '_id', pipeline: [ { '$match': {} }, { '$sort': { createdAt: -1 } }, { '$limit': 11 } ] } }, { '$addFields': { 'values.docs': { '$map': { as: 'doc', in: '$$doc._id', input: '$values.docs' } }, 'values.hasNextPage': { '$gt': [ { '$size': '$values.docs' }, 10 ] } } }, { '$addFields': { 'values.docs': { '$slice': [ '$values.docs', 10 ] } } }, { '$lookup': { as: 'relations.docs', foreignField: 'relatedStock', from: 'stocks', localField: '_id', pipeline: [ { '$match': {} }, { '$sort': { createdAt: -1 } }, { '$limit': 11 } ] } }, { '$addFields': { 'relations.docs': { '$map': { as: 'doc', in: '$$doc._id', input: '$relations.docs' } }, 'relations.hasNextPage': { '$gt': [ { '$size': '$relations.docs' }, 10 ] } } }, { '$addFields': { 'relations.docs': { '$slice': [ '$relations.docs', 10 ] } } }, { '$lookup': { as: 'meetings.docs', foreignField: 'stock', from: 'meetings', localField: '_id', pipeline: [ { '$match': {} }, { '$sort': { createdAt: -1 } }, { '$limit': 11 } ] } }, { '$addFields': { 'meetings.docs': { '$map': { as: 'doc', in: '$$doc._id', input: '$meetings.docs' } }, 'meetings.hasNextPage': { '$gt': [ { '$size': '$meetings.docs' }, 10 ] } } }, { '$addFields': { 'meetings.docs': { '$slice': [ '$meetings.docs', 10 ] } } }, { '$sort': { stockId: 1 } }, { '$count': 'count' } ] } }], {})
`
Is it same bug , or I will open an other bug
